source/adapters/hip/command_buffer.cpp

```diff
@@ -9,156 +9,756 @@
+  }
+
+  try {
+    UR_CHECK_ERROR(hipGraphCreate(&(*phCommandBuffer)->HIPGraph, 0));
+  } catch (...) {
+    return UR_RESULT_ERROR_OUT_OF_RESOURCES;
+  }
+
+  return UR_RESULT_SUCCESS;
 }
 
 UR_APIEXPORT ur_result_t UR_APICALL
-urCommandBufferRetainExp(ur_exp_command_buffer_handle_t) {
-  detail::ur::die("Experimental Command-buffer feature is not "
-                  "implemented for HIP adapter.");
-  return UR_RESULT_ERROR_UNSUPPORTED_FEATURE;
+urCommandBufferRetainExp(ur_exp_command_buffer_handle_t hCommandBuffer) {
+  hCommandBuffer->incrementReferenceCount();
+  return UR_RESULT_SUCCESS;
 }
 
 UR_APIEXPORT ur_result_t UR_APICALL
-urCommandBufferReleaseExp(ur_exp_command_buffer_handle_t) {
-  detail::ur::die("Experimental Command-buffer feature is not "
-                  "implemented for HIP adapter.");
-  return UR_RESULT_ERROR_UNSUPPORTED_FEATURE;
+urCommandBufferReleaseExp(ur_exp_command_buffer_handle_t hCommandBuffer) {
+  if (hCommandBuffer->decrementReferenceCount() != 0)
+    return UR_RESULT_SUCCESS;
+
+  delete hCommandBuffer;
+  return UR_RESULT_SUCCESS;
 }
 
 UR_APIEXPORT ur_result_t UR_APICALL
-urCommandBufferFinalizeExp(ur_exp_command_buffer_handle_t) {
-  detail::ur::die("Experimental Command-buffer feature is not "
-                  "implemented for HIP adapter.");
-  return UR_RESULT_ERROR_UNSUPPORTED_FEATURE;
+urCommandBufferFinalizeExp(ur_exp_command_buffer_handle_t hCommandBuffer) {
+  try {
+    const unsigned long long flags = 0;
+    UR_CHECK_ERROR(hipGraphInstantiateWithFlags(
+        &hCommandBuffer->HIPGraphExec, hCommandBuffer->HIPGraph, flags));
+  } catch (...) {
+    return UR_RESULT_ERROR_UNKNOWN;
+  }
+  return UR_RESULT_SUCCESS;
 }
 
 UR_APIEXPORT ur_result_t UR_APICALL urCommandBufferAppendKernelLaunchExp(
-    ur_exp_command_buffer_handle_t, ur_kernel_handle_t, uint32_t,
-    const size_t *, const size_t *, const size_t *, uint32_t,
-    const ur_exp_command_buffer_sync_point_t *,
-    ur_exp_command_buffer_sync_point_t *) {
-  detail::ur::die("Experimental Command-buffer feature is not "
-                  "implemented for HIP adapter.");
-  return UR_RESULT_ERROR_UNSUPPORTED_FEATURE;
+    ur_exp_command_buffer_handle_t hCommandBuffer, ur_kernel_handle_t hKernel,
+    uint32_t workDim, const size_t *pGlobalWorkOffset,
+    const size_t *pGlobalWorkSize, const size_t *pLocalWorkSize,
+    uint32_t numSyncPointsInWaitList,
+    const ur_exp_command_buffer_sync_point_t *pSyncPointWaitList,
+    ur_exp_command_buffer_sync_point_t *pSyncPoint) {
+  // Preconditions
+  UR_ASSERT(hCommandBuffer->Context == hKernel->getContext(),
+            UR_RESULT_ERROR_INVALID_KERNEL);
+  UR_ASSERT(workDim > 0, UR_RESULT_ERROR_INVALID_WORK_DIMENSION);
+  UR_ASSERT(workDim < 4, UR_RESULT_ERROR_INVALID_WORK_DIMENSION);
+
+  ur_result_t Result = UR_RESULT_SUCCESS;
+  hipGraphNode_t GraphNode;
+
+  std::vector<hipGraphNode_t> DepsList;
+
+  UR_CALL(getNodesFromSyncPoints(hCommandBuffer, numSyncPointsInWaitList,
```

Comment:
This might be a personal preference, but I find accumulation of `Result` introducing unnecessary complexity to the logic. The value is only really important at one place, here: https://github.com/oneapi-src/unified-runtime/blob/b2c7e83521949cd6039cdd1c3874ed0bbd238939/source/adapters/hip/command_buffer.cpp#L266
It would be cleaner to return `Err` in the `catch` and `UR_RESULT_SUCCESS` at the end of the function. 

The same patter of needles result-accumulation is repeated in many functions.

Reply:
The code has been updated to avoid this.